File: go-alert-service/repositories/triggered_alert_repository.go
package repositories

import (
	"database/sql"
	"fmt"

	"iot-alert-service/models"
)

// TriggeredAlertRepository defines the interface for triggered alert data access.
type TriggeredAlertRepository interface {
	GetAll() ([]models.TriggeredAlert, error)
	GetByID(id string) (*models.TriggeredAlert, error)
	Create(ruleID, sensorID string, sensorValue, threshold float64, message string) (*models.TriggeredAlert, error)
	UpdateStatus(id string, status string) (*models.TriggeredAlert, error)
}

// PostgresTriggeredAlertRepository implements TriggeredAlertRepository using Postgres.
type PostgresTriggeredAlertRepository struct {
	db *sql.DB
}

// NewTriggeredAlertRepository creates a new Postgres-backed triggered alert repository.
func NewTriggeredAlertRepository(db *sql.DB) *PostgresTriggeredAlertRepository {
	return &PostgresTriggeredAlertRepository{db: db}
}

// GetAll retrieves all triggered alerts from the database.
func (r *PostgresTriggeredAlertRepository) GetAll() ([]models.TriggeredAlert, error) {
	rows, err := r.db.Query(`
		SELECT id, rule_id, sensor_id, sensor_value, threshold, message, status, created_at, resolved_at
		FROM triggered_alerts ORDER BY created_at DESC
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var alerts []models.TriggeredAlert
	for rows.Next() {
		var alert models.TriggeredAlert
		err := rows.Scan(&alert.ID, &alert.RuleID, &alert.SensorID, &alert.SensorValue,
			&alert.Threshold, &alert.Message, &alert.Status, &alert.CreatedAt, &alert.ResolvedAt)
		if err != nil {
			return nil, err
		}
		alerts = append(alerts, alert)
	}

	if alerts == nil {
		alerts = []models.TriggeredAlert{}
	}

	return alerts, rows.Err()
}

// GetByID retrieves a triggered alert by its ID.
func (r *PostgresTriggeredAlertRepository) GetByID(id string) (*models.TriggeredAlert, error) {
	var alert models.TriggeredAlert
	err := r.db.QueryRow(`
		SELECT id, rule_id, sensor_id, sensor_value, threshold, message, status, created_at, resolved_at
		FROM triggered_alerts WHERE id = $1
	`, id).Scan(&alert.ID, &alert.RuleID, &alert.SensorID, &alert.SensorValue,
		&alert.Threshold, &alert.Message, &alert.Status, &alert.CreatedAt, &alert.ResolvedAt)

	if err == sql.ErrNoRows {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	return &alert, nil
}

// Create inserts a new triggered alert into the database. ID generation uses
// a Postgres SEQUENCE (alert_id_seq), which is atomic and race-free under
// concurrent writes — no explicit transaction needed.
func (r *PostgresTriggeredAlertRepository) Create(ruleID, sensorID string, sensorValue, threshold float64, message string) (*models.TriggeredAlert, error) {
	var nextNum int64
	if err := r.db.QueryRow("SELECT nextval('alert_id_seq')").Scan(&nextNum); err != nil {
		return nil, err
	}
	newID := fmt.Sprintf("alert-%03d", nextNum)

	now := models.Now()

	_, err := r.db.Exec(`
		INSERT INTO triggered_alerts (id, rule_id, sensor_id, sensor_value, threshold, message, status, created_at)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
	`, newID, ruleID, sensorID, sensorValue, threshold, message, "open", now)
	if err != nil {
		return nil, err
	}

	return r.GetByID(newID)
}

// UpdateStatus updates the status of a triggered alert.
func (r *PostgresTriggeredAlertRepository) UpdateStatus(id string, status string) (*models.TriggeredAlert, error) {
	existing, err := r.GetByID(id)
	if err != nil {
		return nil, err
	}
	if existing == nil {
		return nil, nil
	}

	var resolvedAt *string
	if status == "resolved" {
		now := models.Now()
		resolvedAt = &now
	}

	_, err = r.db.Exec(
		"UPDATE triggered_alerts SET status = $1, resolved_at = $2 WHERE id = $3",
		status, resolvedAt, id,
	)
	if err != nil {
		return nil, err
	}

	return r.GetByID(id)
}

File: python-alert-service/repositories/triggered_alert_repository.py
"""Repository for triggered alert data access."""

import sqlite3
from datetime import datetime, timezone
from typing import Optional

from fastapi import Depends

from database import get_db_connection
from models.triggered_alert import AlertStatus, TriggeredAlert


class TriggeredAlertRepository:
    """Data access layer for triggered alerts using SQLite."""

    def __init__(self, db: sqlite3.Connection):
        self.db = db

    def get_all(self) -> list[TriggeredAlert]:
        """Retrieve all triggered alerts."""
        cursor = self.db.execute(
            "SELECT id, rule_id, sensor_id, sensor_value, threshold, message, status, created_at, resolved_at "
            "FROM triggered_alerts ORDER BY created_at DESC"
        )
        rows = cursor.fetchall()
        return [self._row_to_alert(row) for row in rows]

    def get_by_id(self, alert_id: str) -> Optional[TriggeredAlert]:
        """Retrieve a triggered alert by ID."""
        cursor = self.db.execute(
            "SELECT id, rule_id, sensor_id, sensor_value, threshold, message, status, created_at, resolved_at "
            "FROM triggered_alerts WHERE id = ?",
            (alert_id,),
        )
        row = cursor.fetchone()
        return self._row_to_alert(row) if row else None

    def create(
        self,
        rule_id: str,
        sensor_id: str,
        sensor_value: float,
        threshold: float,
        message: str,
    ) -> TriggeredAlert:
        """Create a new triggered alert."""
        cursor = self.db.execute(
            "SELECT MAX(CAST(SUBSTR(id, 7) AS INTEGER)) FROM triggered_alerts WHERE id LIKE 'alert-%'"
        )
        max_num = cursor.fetchone()[0] or 0
        new_id = f"alert-{max_num + 1:03d}"

        now = datetime.now(timezone.utc).isoformat()

        self.db.execute(
            """
            INSERT INTO triggered_alerts (id, rule_id, sensor_id, sensor_value, threshold, message, status, created_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (new_id, rule_id, sensor_id, sensor_value, threshold, message, "open", now),
        )
        self.db.commit()

        return self.get_by_id(new_id)

    def update_status(self, alert_id: str, status: AlertStatus) -> Optional[TriggeredAlert]:
        """Update the status of a triggered alert."""
        existing = self.get_by_id(alert_id)
        if not existing:
            return None

        resolved_at = None
        if status == AlertStatus.RESOLVED:
            resolved_at = datetime.now(timezone.utc).isoformat()

        self.db.execute(
            "UPDATE triggered_alerts SET status = ?, resolved_at = ? WHERE id = ?",
            (status.value, resolved_at, alert_id),
        )
        self.db.commit()

        return self.get_by_id(alert_id)

    def _row_to_alert(self, row: sqlite3.Row) -> TriggeredAlert:
        """Convert a database row to a TriggeredAlert model."""
        return TriggeredAlert(
            id=row["id"],
            rule_id=row["rule_id"],
            sensor_id=row["sensor_id"],
            sensor_value=row["sensor_value"],
            threshold=row["threshold"],
            message=row["message"],
            status=row["status"],
            created_at=row["created_at"],
            resolved_at=row["resolved_at"],
        )


def get_triggered_alert_repository(
    db: sqlite3.Connection = Depends(get_db_connection),
) -> TriggeredAlertRepository:
    """Dependency that provides a TriggeredAlertRepository instance."""
    return TriggeredAlertRepository(db)
